crates/apollo-parser/src/lexer/lookup.rs

```diff
@@ -0,0 +1,100 @@
+use crate::TokenKind;
+
+static PUNCTUATION_CHARS: [Option<TokenKind>; 256] = punctuation_lut();
+static NAMESTART_CHARS: [bool; 256] = namestart_lut();
+
+#[inline]
+pub(crate) fn punctuation_kind(c: char) -> Option<TokenKind> {
+    if c.is_ascii() {
+        PUNCTUATION_CHARS[c as usize]
+    } else {
+        None
+    }
+}
+
+#[inline]
+pub(crate) fn is_namestart(c: char) -> bool {
+    c.is_ascii() && NAMESTART_CHARS[c as usize]
+}
+
+const fn punctuation_lut() -> [Option<TokenKind>; 256] {
+    let mut lut = [None; 256];
+    lut[b'{' as usize] = Some(TokenKind::LCurly);
+    lut[b'}' as usize] = Some(TokenKind::RCurly);
+    lut[b'!' as usize] = Some(TokenKind::Bang);
+    lut[b'$' as usize] = Some(TokenKind::Dollar);
+    lut[b'&' as usize] = Some(TokenKind::Amp);
+    lut[b'(' as usize] = Some(TokenKind::LParen);
+    lut[b')' as usize] = Some(TokenKind::RParen);
+    lut[b':' as usize] = Some(TokenKind::Colon);
+    lut[b',' as usize] = Some(TokenKind::Comma);
+    lut[b'[' as usize] = Some(TokenKind::LBracket);
+    lut[b']' as usize] = Some(TokenKind::RBracket);
+    lut[b'=' as usize] = Some(TokenKind::Eq);
+    lut[b'@' as usize] = Some(TokenKind::At);
+    lut[b'|' as usize] = Some(TokenKind::Pipe);
+
+    lut
+}
+
+/// <https://spec.graphql.org/October2021/#NameStart>
+const fn namestart_lut() -> [bool; 256] {
+    let mut lut = [false; 256];
+    lut[b'a' as usize] = true;
+    lut[b'b' as usize] = true;
+    lut[b'c' as usize] = true;
+    lut[b'd' as usize] = true;
+    lut[b'e' as usize] = true;
+    lut[b'f' as usize] = true;
+    lut[b'g' as usize] = true;
+    lut[b'h' as usize] = true;
+    lut[b'i' as usize] = true;
+    lut[b'j' as usize] = true;
+    lut[b'k' as usize] = true;
+    lut[b'l' as usize] = true;
+    lut[b'm' as usize] = true;
+    lut[b'n' as usize] = true;
+    lut[b'o' as usize] = true;
+    lut[b'p' as usize] = true;
+    lut[b'q' as usize] = true;
+    lut[b'r' as usize] = true;
+    lut[b's' as usize] = true;
+    lut[b't' as usize] = true;
+    lut[b'u' as usize] = true;
+    lut[b'v' as usize] = true;
+    lut[b'w' as usize] = true;
+    lut[b'x' as usize] = true;
+    lut[b'y' as usize] = true;
+    lut[b'z' as usize] = true;
```

Comment:
If only ranged `for` was available in const fn 😁

Reply:
i think this could be a `while` loop

Reply:
you can generate it in a const fn if you dare :p
https://github.com/rust-bakery/parser_benchmarks/blob/master/http/nom-optimized/src/combinators.rs#L194-L242

Reply:
Something like this works:

```rust
const fn generate_table() -> [bool;256] {
    let mut table = [false;256];
    
    let mut i = b'a';
    while i <= b'z' + 1 {
        table[i as usize] = true;
        i += 1;
    }
    
    let mut i = b'A';
    while i <= b'Z' + 1 {
        table[i as usize] = true;
        i += 1;
    }
    
    table
}
```

Seems it's less obvious what's happening and can make the compiler hang if you make a mistake. `const for` would be perfect, but requires unstable features. Also we don't want to make it _too easy_ to make the whole lexer do this, other token types are alot less frequent and benchmarking/profiling shows there was no benefit in those cases.

In fact in some cases (I forgot which) it seemed to actually make things worse

Reply:
I agree writing this stuff out is simpler to understand than a while loop.